src/core.c/Buf.pm6

```diff
@@ -291,23 +291,34 @@ my role Blob[::T = uint8] does Positional[T] does Stringy is repr('VMArray') is
 
     proto method decode(|) {*}
     multi method decode(Blob:D: $encoding = self.encoding // "utf-8") {
-        nqp::p6box_s(
-          nqp::decode(self, Rakudo::Internals.NORMALIZE_ENCODING($encoding))
-        )
+        if Rakudo::Internals.BUILTIN_ENCODINGS{$encoding} {
+            nqp::p6box_s(nqp::decode(self, Rakudo::Internals.NORMALIZE_ENCODING($encoding)))
+        }
+        else { my $decoder = CORE::Encoding::Registry.find($encoding).decoder;
+               $decoder.add-bytes(self);
+               $decoder.consume-all-chars }
     }
 #?if !jvm
     multi method decode(Blob:D: $encoding, Str :$replacement!, Bool:D :$strict = False) {
-        nqp::p6box_s(
-          nqp::decoderepconf(self,
-            Rakudo::Internals.NORMALIZE_ENCODING($encoding),
-            $replacement.defined ?? $replacement !! nqp::null_s(),
-            $strict ?? 0 !! 1))
+        if Rakudo::Internals.BUILTIN_ENCODINGS{$encoding} {
+            nqp::p6box_s( nqp::decoderepconf(self,
+                            Rakudo::Internals.NORMALIZE_ENCODING($encoding),
+                            $replacement.defined ?? $replacement !! nqp::null_s(),
+                            $strict ?? 0 !! 1))
+       }
+       else { my $decoder = CORE::Encoding::Registry.find($encoding).decoder(:$replacement, :$strict);
```

Comment:
Does it always find a decoder?

Reply:
I'm not sure I understand the question.  If the user specifies an encoder that is builtin or that they've added to the registry, then it always finds a decoder.  If they specify a decoder that is neither builtin nor in the registry, then it throws an error `Unknown string encoding: '$encoding'`.  This is the same behavior as before this PR.

(Actually, it's slightly different because the exception is now appropriately typed: it's an `X::Encoding::Unknown.new(name => "non-existant-encoding")`  instead of an `X::AdHoc.new(payload => "Unknown string encoding: 'non-existant-encoding'")` exception.  But it's the message it displays is the same.)